### PR TITLE
Fix performance bug in buildLocationList

### PR DESCRIPTION
**Pull request by tmsri**
In buildLocationList, with basic block sections, we iterate over
every basic block twice to detect section start and end.  This is
sub-optimal and shows up as significantly time consuming when
compiling large functions.

This patch uses the set of sections already stored in MBBSectionRanges
and iterates over sections rather than basic blocks.

When detecting if loclists can be merged, the end label of an entry is
matched with the beginning label of the next entry.  For the section
corresponding to the entry basic block, this is skipped.  This is
because the loc list uses the end label corresponding to the function
whereas the MBBSectionRanges map uses the function end label.

For example:

.Lfunc_begin0:
        .file   <blah>
        .loc    0 4 0                           # ex2.cc:4:0
        .cfi_startproc
.Ltmp0:
        #DEBUG_VALUE: test:i <- 7
        .loc    0 8 5 prologue_end              # ex2.cc:8:5
        ....
.LBB_END0_0:
        .cfi_endproc
        .section        .text._Z4testv,"ax",@progbits,unique,1
...
.Lfunc_end0:
        .size   _Z4testv, .Lfunc_end0-_Z4testv

The debug loc uses ".LBB_END0_0" for the end of the section whereas
MBBSectionRanges uses ".Lfunc_end0".

It is alright to skip this as we already check the section corresponding
to the debugloc entry.

### CODE REVIEW

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 1]
```diff
@@ -1737,6 +1737,12 @@ void AsmPrinter::emitFunctionBody() {
   bool IsEHa = MMI->getModule()->getModuleFlag("eh-asynch");
 
   bool CanDoExtraAnalysis = ORE->allowExtraAnalysis(DEBUG_TYPE);
+  /* Create a slot for the entry basic block section so that the section
+     order is preserved when iterating over MBBSectionRanges. */
+  if (!MF->empty()) {
+    MBBSectionRanges[MF->front().getSectionID()] = MBBSectionRange{CurrentFnBegin, nullptr};
+  }
```
Comment (adrian-prantl) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1744]: Can you update this to LLVM style (// versus /**/ and get rid of the {})?

File: llvm/test/DebugInfo/X86/basic-block-sections_1.ll [hunk 2]
```diff
@@ -16,10 +16,10 @@
 ; NO-SECTIONS: DW_AT_high_pc [DW_FORM_data4] ({{.*}})
 ; BB-SECTIONS: DW_AT_low_pc [DW_FORM_addr] (0x0000000000000000)
 ; BB-SECTIONS-NEXT: DW_AT_ranges [DW_FORM_sec_offset]
+; BB-SECTIONS-NEXT: [{{.*}}) ".text.hot._Z3fooi"
```
Comment (adrian-prantl) [inline llvm/test/DebugInfo/X86/basic-block-sections_1.ll:19]: Why did the order change here?

Comment (tmsri) [reply llvm/test/DebugInfo/X86/basic-block-sections_1.ll:19]: In [here](https://github.com/llvm/llvm-project/blob/488640389faedee8c65566275441779faa620fd2/llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp#L688),

MBBSectionRanges is used to populate DW_AT_ranges.  MBBSectionRanges is a MapVector with deterministic order and I just changed that order to have the entry block's section as the first section in AsmPrinter.cpp.

**Review by jmorse — COMMENTED** [review]
Performance patches most welcome; I can see how stepping through the section-ranges when examining the start-label is going to be faster, and why that necessitates MBBSectionRanges being in the right order.

The change to the end of buildLocationList is less clear however. I believe that section of code is determining whether we should issue a "single-location" location, i.e. saying:

  DW_AT_location      DW_OP_constu 1234

instead of issuing a location list. I would have imagined that this is unrelated to basic-block-sections: if the variable has the same value across its whole scope, surely it doesn't matter which sections the covered instructions are in, and we don't need to issue a location list anyway if that's true.

Specifically, adding the not-function-end qualifier to this clause:

    (RangeIt->second.EndLabel != Asm->getFunctionEnd() &&
         CurEntry->getEndSym() != RangeIt->second.EndLabel)

might disable the detection of whether a variable-location-range over the end block is terminated before the end of the block, and allow the issuing of a "single-location" location. Consider this:

    entry:
        DBG_VALUE 1234
         $rax = something
        ....
    end_bb:
        $rax = something
        DBG_VALUE $noreg
        $rax = something_else
        ret

Here the variable location covers the whole function except the last instruction -- perhaps there's a late variable assignment that's been optimised away. In this situation we want to issue a location list that covers the whole function except the final instruction, but I believe with this patch (and that extra clause of `RangeIt->second.EndLabel != Asm->getFunctionEnd()`) we'll instead emit a single-location location. That can then lead to a stale/out-of-date variable value being presented to the user on the final instruction of the program.

I might have missed something subtle about the labels though (I'm not familiar with basic-block sections, and it seems fiddly).

**Comment by tmsri** [comment]
> Here the variable location covers the whole function except the last instruction -- perhaps there's a late variable assignment that's been optimised away. In this situation we want to issue a location list that covers the whole function except the final instruction, but I believe with this patch (and that extra clause of `RangeIt->second.EndLabel != Asm->getFunctionEnd()`) we'll instead emit a single-location location. That can then lead to a stale/out-of-date variable value being presented to the user on the final instruction of the program.
> 
> I might have missed something subtle about the labels though (I'm not familiar with basic-block sections, and it seems fiddly).

I did think of the case where the .loc only spans some subset of the entry section and there is another variable assignment that will cover the remaining subset of the entry section.  I think you are referring to this one, I am not sure what you mean by "optimised away" though, like there is an assignment but the .loc is missing? 

In this case, the values won't match between the first loc and the subsequent loc and the coalescing will not happen.

I will add an explicit test to cover this.  I understand your concern around that check, but I am unable to find a counter example that could violate it.   Let me refresh the patch with the new test case.  Thanks for taking a look!

**Comment by tmsri** [comment]
An accidental rebase from my end messed up this PR, apologies.  Let me fix that, please ignore this PR until then!

**Pull request closed without merging**